lnst/Recipes/ENRT/PerfTestMixins/DevFlowsPinningMixin.py

```diff
@@ -0,0 +1,249 @@
+
+        E.g. if :attr:`flow.steer_by` is "src-ip", it returns
+        :attr:`flow.generator_bind` IP address.
+
+        :param flow: flow object to pin
+        :returns: steering key
+        """
+        if steer_by == "src-ip":
+            return flow.generator_bind
+        elif steer_by == "dst-ip":
+            return flow.receiver_bind
+        elif steer_by == "src-port":
+            return flow.generator_port
+        elif steer_by == "dst-port":
+            return flow.receiver_port
+        else:
+            raise NotImplementedError(
+                f"Flow steering by {steer_by} not implemented (yet)."
+            )
+
+    def _get_flow_type(self, flow: Flow, steer_by) -> str:
+        """
+        Returns flow protocol based on which steering is done.
+        E.g.
+            if steering by IP, it returns "ip4" or "ip6".
+            if steering by TCP, it returns "tcp4" or "tcp6".
+
+        :param flow: flow object to pin
+        :returns: flow protocol+IP version
+        """
+        ip_version = (
+            "4" if isinstance(ipaddress(flow.receiver_bind), Ip4Address) else "6"
+        )
+
+        protocol = None
+        if "ip" in steer_by:  # e.g. src-ip, dst-ip
+            protocol = "ip"
+
+        elif "port" in steer_by:  # e.g. src-port, dst-port
+            if "tcp" in flow.type:
+                protocol = "tcp"
+            elif "sctp" in flow.type:
+                protocol = "sctp"
+            elif "udp" in flow.type or "xdp" in flow.type:  # xdp tests uses udp
+                protocol = "udp"
+            else:
+                raise NotImplementedError(
+                    f"Flow steering for {flow.type} not implemented (yet)."
+                )
+        else:
+            raise NotImplementedError(
+                f"Flow steering by {steer_by} not implemented (yet)."
+            )
+
+        return protocol + ip_version
+
+    def remove_perf_test_tweak(self, perf_config):
+        if "flow_steering_rules" in perf_config.perf_test_tweak_config:
+            for nic, rules in perf_config.perf_test_tweak_config[
+                "flow_steering_rules"
+            ].items():
+                for id, _ in rules:
+                    nic.netns.run(f"ethtool -N {nic.name} delete {id}")
+                    # ^ remove steering rule
+
+        super().remove_perf_test_tweak(perf_config)
+
+    def steer_flow_by(
+        self, flow
+    ) -> Optional[Literal["dst-ip", "src-ip", "dst-port", "src-port"]]:
+        """
+        This method should return an steering key, based on
+        which the flow should be steered across the queues.
+        E.g. if steering by src-ip, it returns "src-ip" and
+        the flow will be steered based on its' source IP address.
+
+        If None is returned, flow won't be steered.
+
+        :param flow: flow object to pin
+        :returns: steering key
+        """
+        return getattr(self.params, "steer_flows_by", None)
```

Comment:
is this enough? the function has a `flow` parameter which isn't used. or is the parameter supposed to be a single value that is used for all flows?